nixos/modules/services/misc/jellyfin.nix

```diff
@@ -0,0 +1,59 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.jellyfin;
+in
+{
+  options = {
+    services.jellyfin = {
+      enable = mkEnableOption "Jellyfin Media Server";
+
+      user = mkOption {
+        type = types.str;
+        default = "jellyfin";
+        description = "User account under which Jellyfin runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "jellyfin";
+        description = "Group under which jellyfin runs.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.jellyfin = {
+      description = "Jellyfin Media Server";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = rec {
+        Type = "simple";
+        User = cfg.user;
+        Group = cfg.group;
+        StateDirectory = "jellyfin";
+        CacheDirectory = "jellyfin";
+        ExecStart = "${pkgs.jellyfin}/bin/jellyfin --datadir '/var/lib/${StateDirectory}' --cachedir '/var/cache/${CacheDirectory}'";
+        Restart = "on-failure";
+      };
+    };
+
+    users.users = mkIf (cfg.user == "jellyfin") {
+      jellyfin = {
+        group = cfg.group;
+        uid = config.ids.uids.jellyfin;
```

Comment:
Why not just create the user and *not* specify the `uid`?

Reply:
If the UID/GID changes across rebuilds, the data/cache directory won't be readable by Jellyfin

Reply:
I also assumed that was true until recently. See https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/update-users-groups.pl#L128 for an explanation.

Reply:
also Jellyfin (like Emby) can create files in the library folders like NFO and pictures

Reply:
The service will still function properly without having an explicit `uid` or `gid` assigned.

Reply:
Correct me if I'm wrong, but from the perl script you have linked @aanderse NixOS will keep compare UIDs and GIDs after and before a rebuild and force them to be the same if they change, right? But it prints a warning message which makes me think that we should probably avoid triggering it, but I don't know enough about the code convention of nixpkgs in this particular instance ^^

Reply:
@Infinisil I know you have [mentioned](https://github.com/NixOS/nixpkgs/pull/56423#discussion_r263802253) this at least once or twice... Care to weigh in?

Reply:
@minijackson The warning will only appear if the uid/gid tries to be changed, which won't happen under normal circumstances.

If people are worried about the service becoming unfunctional due to permissions after a reinstall (which btw will only happen if you forget to restore /var/lib/nixos, which you should do because of this anyways), you can make the service fix the permissions on startup if they're not correct, which [`StateDirectory` does already](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectory=):

> If the specified directories already exist and their owning user or group do not match the configured ones, all files and directories below the specified directories as well as the directories themselves will have their file ownership recursively changed to match what is configured.

Static uid/gid's should be avoided because if you look a bit further down in `ids.nix`, you'll see this little comment:

https://github.com/NixOS/nixpkgs/blob/ed569c149267f7ab3e508de92939257d29bd2dfe/nixos/modules/misc/ids.nix#L340-L343

So we have about 85 static uid/gid's left because everybody's been adding them here over the past years. In *some* circumstances like services that span across multiple machines and mount filesystems or whatever, static uid/gid's might indeed be needed, so we should keep these available for such services.

In theory we could extend this range, but then we'd have to use id's above 30000 or so (if possible at all), because everything else is already reserved, see https://github.com/NixOS/nixpkgs/blob/a0bb3dd8ac4fb1e0c5fa4b897b97b7b5e79a5f31/nixos/modules/programs/shadow.nix#L13-L16